src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,289 @@
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_ACK
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection ack sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_recv_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+
+    ucs_assertv((status != UCS_OK) ||
+                (conn_pkt.event == UCT_TCP_CM_CONN_ACK), "ep=%p", ep);
+    return status;
+}
+
+unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
+{
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
     ucs_status_t status;
 
     status = ucs_socket_connect_nb_get_status(ep->fd);
-    if (status == UCS_INPROGRESS) {
-        return 0;
-    } else if (status != UCS_OK) {
+    if (ucs_unlikely(status != UCS_OK)) {
+        if (status == UCS_INPROGRESS) {
+            return 0;
+        }
         goto err;
     }
 
-    iface->outstanding--;
+    status = uct_tcp_cm_send_conn_req(ep);
+    if (ucs_unlikely(status != UCS_OK)) {
+        ucs_debug("tcp_ep %p: unable to send connection request to %s",
+                  ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                       UCS_SOCKADDR_STRING_LEN));
+        return 0;
+    }
 
-    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAITING_ACK);
+    uct_tcp_ep_mod_events(ep, EPOLLIN, EPOLLOUT);
 
     ucs_assertv((ep->tx.length == 0) && (ep->tx.offset == 0) &&
                 (ep->tx.buf == NULL), "ep=%p", ep);
-
-    return uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
+    return 1;
 
 err:
-    iface->outstanding--;
     uct_tcp_ep_set_failed(ep);
     return 0;
 }
 
-void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep, uct_tcp_ep_conn_state_t new_conn_state)
+unsigned uct_tcp_cm_conn_ack_rx_progress(uct_tcp_ep_t *ep)
 {
-    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_tcp_iface_t);
-    char str_local_addr[UCS_SOCKADDR_STRING_LEN], str_remote_addr[UCS_SOCKADDR_STRING_LEN];
-    uct_tcp_ep_conn_state_t old_conn_state;
+    unsigned ret = 0;
+    ucs_status_t status;
 
-    old_conn_state = ep->conn_state;
-    ep->conn_state = new_conn_state;
+    status = uct_tcp_cm_recv_conn_ack(ep);
+    if (ucs_unlikely(status != UCS_OK)) {
+        uct_tcp_ep_mod_events(ep, 0, EPOLLIN);
+        return ret;
+    }
+
+    ret++;
```

Comment:
why need this variable?
in first return, just return 0
in last return, return 1 + ...
similarly for uct_tcp_cm_conn_req_rx_progress